### PR TITLE
Add option for SBT creation to be localized 

### PR DESCRIPTION
**Pull request by olgabot**
Adds a new `LocalizedSBT` class in `sbtmh.py` that adds new `SourmashSignature` leaves into their optimal position, sharing a parent with the leaf with which it shares the highest similarity. This enables building a Nearest Neighbor graph (https://github.com/dib-lab/sourmash/pull/710) directly from the `LocalizedSBT`, which can then be used for graph-based/community clustering and UMAP.

See also #756 

- [x] Is it mergeable?
- [x] `make test` Did it pass the tests?
- [ ] `make coverage` Is the new code covered?
- [ ] Did it change the command-line interface? Only additions are allowed
  without a major version increment. Changing file formats also requires a
  major version number increment.
- [ ] Was a spellchecker run on the source code and documentation after
  changes were made?

### CODE REVIEW

**Comment by ctb** [comment]
hi olga, you need to run a clean rust build, I think.

**Comment by ctb** [comment]
Right, I think there are stale build files, that wouldn't be affected by
new environment etc. Try `make clean` first?

**Comment by ctb** [comment]
On Tue, Mar 31, 2020 at 07:33:15AM -0700, Olga Botvinnik wrote:
> Ohh okay I didn't realize "clean rust build" == `rustup update && make clean`

sorry - I didn't have the commands handy myself :)

`make clean` is a good default thing to do and it has worked for me
in the past with similar errors.

**Comment by ctb** [comment]
On Tue, Mar 31, 2020 at 07:34:55AM -0700, Olga Botvinnik wrote:
> THANK YOU!! My tests are *correctly* not passing now :)

lol!

please don't close this issue - we should add this error message and
`make clean` to the docs.

**Comment by luizirber** [comment]
> What is the reasoning behind the sourmash bioconda recipe not specifying a Rust version? That seems like a good place to link to conda-forge built binaries.

sourmash has a build-time dependency on Rust (to build the shared library), but not during runtime (for Python it looks like any C extension).

It makes a lot of sense for a dev environment, but that's not the primary use case, nor something that conda supports...

**Comment by ctb** [comment]
hi @phoenixAja @olgabot could you put some text in the description of this issue? I had to dig into the source code to figure out what it was about :) thx!

**Comment by luizirber** [comment]
This is looking very good!

Any reason you decided to do a new SBT instead of changing the current one?

I built one with 141k signatures, it used 37.9 GB to build and 44 minutes. That's in the same ballpark as the current SBT. 
I think this should also be the default on `index`, and only use the regular SBT if `d != 2` (which I don't think anyone uses, BTW).

I'm also running a long query, and this does wonders to memory consumption (it's stable around 3.3 GB). Still takes longer than the LCA to run, but that's due to other limitations unrelated to localization.

File: sourmash/sbtmh.py [hunk 1]
```diff
@@ -77,6 +80,207 @@ def data(self, new_data):
         self._data = new_data
 
 
+class LocalizedSBT(SBT):
+    """A Sequence Bloom Tree implementation which guarantees new leaves are plaaced in
```
Comment (ctb) [inline sourmash/sbtmh.py]: Please expand/fix :)

File: sourmash/sbtmh.py [hunk 2]
```diff
@@ -77,6 +80,207 @@ def data(self, new_data):
...
+        if not self._nodes:
+            self.next_node = 1
+            return 0
+
+        if not self._leaves:
+            self.next_node = 2
+            return 1
+        # Not an empty tree, can search
+
+        # TODO: There is probably a way better way to write this logic - @olgabot
```
Comment (ctb) [inline sourmash/sbtmh.py]: Right, I would like to suggest that the code below be refactored into smaller functions. It's already pretty clear (good job w/comments and formatting!!) but it's complex enough to warrant more :). I also want to place a marker that this code should also have a lot of associated tests, perhaps at the unit test level. We have a lot of functional tests on SBTs, so I'm reasonably sure the code works, but it would be good to get some edge cases and so on at the unit test level in the code.

File: sourmash/sbt.py [hunk 3]
```diff
@@ -187,43 +193,46 @@ def add_node(self, node):
...
+            self._rebuild_node(parent.pos)
+            node.update(self._nodes[parent.pos])
+            parent = self.parent(parent.pos)
+
+    def insert_new_internal_node_with_children(self, node, parent):
+        # Create a new internal node
+        # node and parent are children of new internal node
+        n = Node(self.factory, name="internal." + str(parent.pos))
+        self._nodes[parent.pos] = n
+        c1, c2 = self.children(parent.pos)[:2]
```
Comment (phoenixAja) [inline sourmash/sbt.py:256]: would this also hold if the tree had more than two children per node? Or are we only considering at most two children per node?

Comment (olgabot) [reply sourmash/sbt.py:256]: Yeah probably wouldn't hold :) This implementation throws a `NotmplementedError` if someone specifies `n_children = d != 2` for now.

File: sourmash/sbt.py [hunk 4]
```diff
@@ -254,7 +263,11 @@ def find(self, search_fn, *args, **kwargs):
 
                     # leaf node? it's a match!
                     if isinstance(node_g, Leaf):
-                        matches.append(node_g)
+                        # Return tuple of node data and position
+                        if return_pos:
+                            matches.append((node_g, node_p))
```
Comment (phoenixAja) [inline sourmash/sbt.py:311]: I think a named tuple/ dataclass here would be nice, especially because I could see how someone could easily confuse `node_g` and `node_p` variables, but it could be overkill depending on how often `matches` is used by the end user: 

maybe something like 
`MatchedNode(data=node_g, position=node_p)`

File: sourmash/sbt.py [hunk 5]
```diff
@@ -279,13 +292,17 @@ def search(self, query, *args, **kwargs):
         best_only = kwargs['best_only']
         unload_data = kwargs.get('unload_data', False)
 
+        # If True, returns the leaf node (rather than the signature data) and the
+        # position of the leaf on the tree
```
Comment (phoenixAja) [inline sourmash/sbt.py:361]: awesome comments!!

File: sourmash/sbt.py [hunk 6]
```diff
@@ -279,13 +292,17 @@ def search(self, query, *args, **kwargs):
...
+        return_leaf = kwargs.get('return_leaf', False)
```
Comment (phoenixAja) [inline sourmash/sbt.py]: you know this code base better than me, but if you're querying the kwargs a lot like this do you think this whole method might be better adapted to it's own class? Maybe named `Finder` or something like that? That way you could also break up this function a little more and create methods for searching, finding scaled value of the tree, creating a search queue, etc?

File: sourmash/sbtmh.py [hunk 7]
```diff
@@ -77,6 +80,207 @@ def data(self, new_data):
...
+    -----
+    We use two dicts to store the tree structure: One for the internal nodes,
+    and another for the leaves (datasets).
+    """
+
+    def __init__(self, factory, d=2, storage=None, track_abundance=False,
+                 do_containment=False):
+        if d != 2:
+            raise NotImplementedError("LocalizedSBT is only implemented for when the "
+                                      "number of children is 2, d=2")
```
Comment (phoenixAja) [inline sourmash/sbtmh.py]: ah ok this answers my earlier question!

**Comment by ctb** [comment]
On Mon, Apr 06, 2020 at 09:05:50AM -0700, Phoenix Logan wrote:
> @phoenixAja commented on this pull request.
> 
> > @@ -279,13 +292,17 @@ def search(self, query, *args, **kwargs):
>          best_only = kwargs['best_only']
>          unload_data = kwargs.get('unload_data', False)
>  
> +        # If True, returns the leaf node (rather than the signature data) and the
> +        # position of the leaf on the tree
> +        return_leaf = kwargs.get('return_leaf', False)
> 
> you know this code base better than me, but if you're querying the kwargs a lot like this do you think this whole function might be better adapted to a class? Maybe named `Finder` or something like that? That way you could also break up this function a little more and create methods for searching, finding scaled value of the tree, creating a search queue, etc? 

I like the idea but I propose that it be a separate PR (either into this one,
or into master) so it can be reviewed separately.

File: sourmash/sbtmh.py [hunk 8]
```diff
@@ -77,6 +80,207 @@ def data(self, new_data):
...
+        SBT.__init__(self, factory=factory, d=d, storage=storage)
```
Comment (phoenixAja) [inline sourmash/sbtmh.py]: I may not be understanding correctly, (not sure why you're passing in `self` here) but I think this would be easier to maintain if you called `super().__init__` here instead of `SBT.__init__` i think this is a more common pattern in python. But then again i could be misunderstanding the use case of the parent class 
```suggestion
        super().__init__(factory=factory, d=d, storage=storage)
```

File: sourmash/sbt.py (hunk 3 above)
Comment (phoenixAja) [reply sourmash/sbt.py:256]: @olgabot i just saw in the localized implementation! Sorry :)

File: sourmash/sbtmh.py (hunk 2 above)
Comment (phoenixAja) [reply sourmash/sbtmh.py]: i think it would be cool to incorporate some factory methods to contain a fair amount of the if/else logic.

File: sourmash/sbtmh.py [hunk 9]
```diff
@@ -77,6 +85,210 @@ def data(self, new_data):
...
+                    if new_leaf_similarity > child_similarity:
+                        # New leaf is *more* similar than the existing child
+                        # --> displace existing child
+
+                        # Get this child's displaced position
+                        displaced_position = other_child.pos
+
+                        # Place the less similar child in the neighboring node
+                        grandparent = self.parent(most_similar_parent.pos)
+                        parent_sibling = [x for x in self.children(grandparent.pos)
```
Comment (phoenixAja) [inline sourmash/sbtmh.py]: `grandparent.pos` made me laugh 😄

File: sourmash/sbtmh.py [hunk 10]
```diff
@@ -77,6 +85,210 @@ def data(self, new_data):
...
+        while parent:
+            self._rebuild_node(parent.pos)
+            node.update(self._nodes[parent.pos])
+            parent = self.parent(parent.pos)
+
+    def relocate_children_to_new_internal_node(self, grandparent, node, parent,
+                                               parent_sibling):
+        new_internal_node = Node(self.factory, name="internal." + str(parent.pos))
+        self._nodes[parent.pos] = new_internal_node
+        c1, c2 = self.children(parent.pos)[:2]
```
Comment (phoenixAja) [inline sourmash/sbtmh.py]: could you unpack this instead of index?

```suggestion
        c1, c2 = *self.children(parent.pos)
```

File: tests/test_sbtmh.py [hunk 11]
```diff
@@ -0,0 +1,118 @@
...
+    #      1           2
+    #    /   \       /   \
+    # a: 3  c: 4   d: 5  b: 6
+    # Tree: (track_abundance=False)
+    #             0
+    #         /      \
+    #      1           2
+    #    /   \       /   \
+    # a: 3  b: 4   c: 5  d: 6
+
```
Comment (phoenixAja) [inline tests/test_sbtmh.py:110]: nice comments! This visualization is really cool for understanding what's going on 👍

**Comment by phoenixAja** [comment]
> On Mon, Apr 06, 2020 at 09:05:50AM -0700, Phoenix Logan wrote: @phoenixAja commented on this pull request. > @@ -279,13 +292,17 @@ def search(self, query, *args, **kwargs): best_only = kwargs['best_only'] unload_data = kwargs.get('unload_data', False) + # If True, returns the leaf node (rather than the signature data) and the + # position of the leaf on the tree + return_leaf = kwargs.get('return_leaf', False) you know this code base better than me, but if you're querying the kwargs a lot like this do you think this whole function might be better adapted to a class? Maybe named `Finder` or something like that? That way you could also break up this function a little more and create methods for searching, finding scaled value of the tree, creating a search queue, etc?
> I like the idea but I propose that it be a separate PR (either into this one, or into master) so it can be reviewed separately.

Yes for sure! I can see where that would be out of scope for this

**Comment by luizirber** [comment]
>     2\. Now that I think about it, it may also be useful to have a command for `sourmash reindex` to run this on existing indices. Another PR perhaps?

Can't do a deeper review now, but I really like the `sourmash reindex` idea! I was thinking about adding more flags to `sourmash index` (or even making it a subcommand...), but both approaches get increasingly gnarly over time. `reindex` would also serve as a good entry point from converting from SBT -> LCA and vice versa, or any future indices.

**Comment by ctb** [comment]
hi @olgabot let us know when things are ready for re-review. No hurry, just wanted to make sure I wasn't missing anything :)

File: sourmash/sbtmh.py [hunk 12]
```diff
@@ -14,10 +16,16 @@ def load_sbt_index(filename, print_version_warning=True):
...
 
 
-def create_sbt_index(bloom_filter_size=1e5, n_children=2):
+def create_sbt_index(bloom_filter_size=1e5, n_children=2, not_localized=False):
     "Create an empty SBT index."
     factory = GraphFactory(1, bloom_filter_size, 4)
-    tree = SBT(factory, d=n_children)
+    if not_localized or n_children != 2:
+        if n_children != 2:
+            notify("n_children != 2 in tree, defaulting to non-localized sequence "
```
Comment (ctb) [inline sourmash/sbtmh.py]: I think this should raise a `ValueError` instead, since this is an unlikely situation for anyone using the command line to run into!

File: sourmash/sbtmh.py [hunk 13]
```diff
@@ -77,6 +85,256 @@ def data(self, new_data):
...
+        if d != 2:
+            raise NotImplementedError("LocalizedSBT is only implemented for when the "
+                                      "number of children is 2, d=2")
+        super().__init__(factory=factory, d=d, storage=storage)
+        self.track_abundance = track_abundance
+        self.ignore_abundance = not self.track_abundance
+        self.do_containment = do_containment
+
+    def find_most_similar_leaf(self, node):
+            search_results = self.search(
```
Comment (ctb) [inline sourmash/sbtmh.py]: wonky indentation!

File: sourmash/cli/index.py [hunk 14]
```diff
@@ -44,6 +44,13 @@ def subparser(subparsers):
         '--scaled', metavar='FLOAT', type=float, default=0,
         help='downsample signatures to the specified scaled factor'
     )
+    subparser.add_argument(
+        '--not-localized', action='store_true', default=False,
+        help='Do not create a localized SBT index. A localized index guarantees that '
+             'any two leaves sharing a parent are more similar to leaves not sharing '
+             'a parent. Localized indices are required for building a nearest neighbor'
+             ' graph from the SBT.'
+    )
```
Comment (luizirber) [inline sourmash/cli/index.py:84]: I don't think this option is necessary: if `d != 2` then build the current `SBT`, otherwise always build the `LocalizedSBT`.

File: sourmash/commands.py [hunk 15]
```diff
@@ -338,7 +338,8 @@ def index(args):
     if args.append:
         tree = load_sbt_index(args.sbt_name)
     else:
-        tree = create_sbt_index(args.bf_size, n_children=args.n_children)
+        tree = create_sbt_index(args.bf_size, n_children=args.n_children,
+                                not_localized=args.not_localized)
```
Comment (luizirber) [inline sourmash/commands.py]: This can be decided inside `create_sbt_index` by checking the `n_children` value, so I think it's fine to revert.

```suggestion
        tree = create_sbt_index(args.bf_size, n_children=args.n_children)
```

File: sourmash/sbtmh.py [hunk 16]
```diff
@@ -14,10 +16,16 @@ def load_sbt_index(filename, print_version_warning=True):
...
     "Create an empty SBT index."
     factory = GraphFactory(1, bloom_filter_size, 4)
-    tree = SBT(factory, d=n_children)
+    if not_localized or n_children != 2:
+        if n_children != 2:
+            notify("n_children != 2 in tree, defaulting to non-localized sequence "
+                   "bloom tree (SBT) index")
+        tree = SBT(factory, d=n_children)
+    else:
+        tree = LocalizedSBT(factory, d=n_children)
```
Comment (luizirber) [inline sourmash/sbtmh.py]:
```suggestion
    if n_children != 2:
        tree = SBT(factory, d=n_children)
    else:
        tree = LocalizedSBT(factory, d=n_children)
```

File: sourmash/sbtmh.py [hunk 17]
```diff
@@ -14,10 +16,16 @@ def load_sbt_index(filename, print_version_warning=True):
                     print_version_warning=print_version_warning)
 
 
-def create_sbt_index(bloom_filter_size=1e5, n_children=2):
+def create_sbt_index(bloom_filter_size=1e5, n_children=2, not_localized=False):
```
Comment (luizirber) [inline sourmash/sbtmh.py]:
```suggestion
def create_sbt_index(bloom_filter_size=1e5, n_children=2):
```

**Comment by ctb** [comment]
no worries, hope your head reattaches?

File: sourmash/sbt.py [hunk 18]
```diff
@@ -121,18 +122,27 @@ class SBT(Index):
...
         self._leaves = {}
         self.d = d
         self.next_node = 0
         self.storage = storage
 
+    @property
+    def _missing_nodes(self):
+        return {i for i in range(max(self._nodes))
+                if i not in self._nodes and i not in self._leaves}
+
```
Comment (luizirber) [inline sourmash/sbt.py]: This doesn't quite work, because `self._nodes` might be empty. (It worked during loading because then we had `max_node` to keep track.)

An implementation that doesn't assume dense SBTs is below:
```suggestion
     # TODO: make it into a cached_property, clean it if new node is inserted                                                                     
     @property
     def _missing_nodes(self):
         all_nodes = set()
         current_level = set()

         for i in self._leaves:
             parent = self.parent(i)
             if parent is not None:
                 current_level.add(parent.pos)

         while current_level:
             previous_level = set()
             for i in current_level:
                 parent = self.parent(i)
                 if parent is not None:
                     previous_level.add(parent.pos)
             all_nodes.update(current_level)
             current_level = previous_level

         return all_nodes - set(self._nodes)
```

The issue is that it might be expensive to calculate this all the time, so probably the [`@cached_property`](https://github.com/dib-lab/sourmash/blob/baaa3107ae74ccbeb1838c8f1589cd7dfa66b0c7/sourmash/lca/lca_db.py#L15) decorator from the LCA db might be useful here too.

(and `@cached_property` should be in some other place in sourmash, because it is not LCA DB-specific =])

**Comment by luizirber** [comment]
> At first, I thought of doing this with growing list of `self._ignore_leaves_when_searching` attribute on the SBT, but appending to a Python list thousands of times isn't great. 

It is also not so bad, because Python lists allocate extra capacity when they grow, so inserting is amortized (it doesn't need to reallocate all the time). And a `set` might be a good option too, because then you don't need to worry about duplicates.

> That would make the total worst-case time of inserting a new leaf to be `O(n)` which seems really big, and seems to also defeat the purpose of trees a bit...

I think this depends on use cases. If you're building an index to use it once, it is probably too costly. But most users don't build a new index at all, and use one of the prepared databases for genbank and refseq, or create their own and use it repeatedly. For these paying the cost of building a more efficient index for search/gather is a very good tradeoff.

The implementation changed since I tried this PR, but at that time the build time was not much larger than for a regular SBT. And while your analysis is correct on the complexity, I don't think we will be hitting the worst case all the time (which is what matters for our practical purposes =])

Finally, I appreciate worrying about performance, but don't worry to much about getting the perfect implementation. We survived with a very naive insertion method until now, and we can make this one better over time too =]

**Comment by ctb** [comment]
Hi @olgabot, part of any review will be looking at the tests you changed,
so I wouldn't worry about that - just get it working to your satisfaction
and we'll iterate during a review!

**Comment by ctb** [comment]
huh! I'll take a look.

File: sourmash/commands.py [hunk 19]
```diff
@@ -387,32 +391,35 @@ def index(args):
...
         sys.exit(-1)
+    if return_n:
+        return tree, n
+    else:
+        return tree
 
-    notify('loaded {} sigs; saving SBT under "{}"', n, args.sbt_name)
-    tree.save(args.sbt_name, sparseness=args.sparseness)
+
+def check_signature_compatibilty_to_tree(ksizes, moltypes, nums, scaleds):
```
Comment (ctb) [inline sourmash/commands.py:420]: in sourmash_args, we have `check_tree_is_compatible`; they do slightly different things and I think it's fine to use this new function, but we should name it something else.

Comment (ctb) [reply sourmash/commands.py:420]: (maybe prefix it with an _ to indicate that it's a helper function that shouldn't be used outside this module?)

File: sourmash/lca/lca_db.py [hunk 20]
```diff
@@ -4,27 +4,11 @@
 import json
 import gzip
 from collections import OrderedDict, defaultdict, Counter
-import functools
 
-import sourmash
 from sourmash._minhash import get_max_hash_for_scaled
-from sourmash.logging import notify, error, debug
+from sourmash.logging import notify, debug
```
Comment (ctb) [inline sourmash/lca/lca_db.py:9]: it looks to me like you're cleaning up unnecessary imports here, yes? that's great! but it's a little confusing to do that in an unrelated PR and trips up code review - I was like, why are they changing lca_db!?

Comment (ctb) [reply sourmash/lca/lca_db.py:9]: oh, I see, you had to touch this file to move `cached_property` to utils. never mind!

**Comment by ctb** [comment]
> I'm also very confused because "fixing" the tests was mostly changing a bunch of output results for `gather` and `search`. I'm not 100% certain why they are changing with a localized SBT so I welcome feedback on whether those should actually be changed.

hi @olgabot,

I took at look at `test_sourmash.py:test_gather_metagenome`, which was one of the tests that changed. The test builds a new SBT using `tests/test-data/gather/GCF*.sig` and then does `sourmash gather` with `gather tests/test-data/gather/combined.sig` against it, after setting `--threshold-bp=0`.

I ran into two problems which suggests to me that there's a bug in this PR somewhere --

first, I compared the results above with not using an SBT by running:

```
sourmash gather tests/test-data/gather/combined.sig tests/test-data/gather/GCF*.sig -k 21 --threshold-bp=0
```

and I got 12 matches total, matching what was in the original test. Since the results of running any search on an index should be the same as running a search on the signatures, the original test is correct. (whew :)

second, I stumbled across some order dependence on the input signatures. If I do

```
sourmash index -k 21 test1 tests/test-data/gather/GCF_000011885.1_ASM1188v1_genomic.fna.gz.sig tests/test-data/gather/GCF_000016785.1_ASM1678v1_genomic.fna.gz.sig tests/test-data/gather/GCF_000006945.2_ASM694v2_genomic.fna.gz.sig tests/test-data/gather/GCF_000009085.1_ASM908v1_genomic.fna.gz.sig tests/test-data/gather/GCF_000016045.1_ASM1604v1_genomic.fna.gz.sig tests/test-data/gather/GCF_000007545.1_ASM754v1_genomic.fna.gz.sig tests/test-data/gather/GCF_000195995.1_ASM19599v1_genomic.fna.gz.sig tests/test-data/gather/GCF_000008545.1_ASM854v1_genomic.fna.gz.sig tests/test-data/gather/GCF_000009505.1_ASM950v1_genomic.fna.gz.sig tests/test-data/gather/GCF_000009525.1_ASM952v1_genomic.fna.gz.sig tests/test-data/gather/GCF_000008105.1_ASM810v1_genomic.fna.gz.sig tests/test-data/gather/GCF_000018945.1_ASM1894v1_genomic.fna.gz.sig
sourmash gather tests/test-data/gather/combined.sig test1 --threshold-bp=0
```
I get different 11 results; if I index the signatures in a different order,
```
sourmash index -k 21 test2 tests/test-data/gather/GCF_000006945.2_ASM694v2_genomic.fna.gz.sig tests/test-data/gather/GCF_000007545.1_ASM754v1_genomic.fna.gz.sig tests/test-data/gather/GCF_000008105.1_ASM810v1_genomic.fna.gz.sig tests/test-data/gather/GCF_000008545.1_ASM854v1_genomic.fna.gz.sig tests/test-data/gather/GCF_000009085.1_ASM908v1_genomic.fna.gz.sig tests/test-data/gather/GCF_000009505.1_ASM950v1_genomic.fna.gz.sig tests/test-data/gather/GCF_000009525.1_ASM952v1_genomic.fna.gz.sig tests/test-data/gather/GCF_000011885.1_ASM1188v1_genomic.fna.gz.sig tests/test-data/gather/GCF_000016045.1_ASM1604v1_genomic.fna.gz.sig tests/test-data/gather/GCF_000016785.1_ASM1678v1_genomic.fna.gz.sig tests/test-data/gather/GCF_000018945.1_ASM1894v1_genomic.fna.gz.sig tests/test-data/gather/GCF_000195995.1_ASM19599v1_genomic.fna.gz.sig
sourmash  gather tests/test-data/gather/combined.sig test2 --threshold-bp=0
```
I get 8 results.

Interestingly these results all seem to hold if I specify `--not-localized`, too. Not sure how to interpret that, but maybe some of the refactoring is the problem, rather than the addition of localization?

File: sourmash/cli/index.py (hunk 14 above)
Comment (ctb) [reply sourmash/cli/index.py:84]: well, it's nice for debugging :)

**Comment by luizirber** [comment]
I think there might be a corner case that was missed... I remember checking this some time ago, but didn't dig further. This is what I was changing at the time: https://github.com/dib-lab/sourmash/commit/0d68bf8fb9dab64237d84fc9cd5873c0c7149d55

I updated the comment/diagram in [this test](https://github.com/dib-lab/sourmash/commit/0d68bf8fb9dab64237d84fc9cd5873c0c7149d55#diff-e063f81aa1b9b97bcd28e63aa12cf863L216) to reflect it, but it was related to pushing a new level too soon.

**Comment by ctb** [comment]
I did some digging! tl;dr at least one obvious problem, not sure what the cause is.

code for SBT print stuff is here: https://github.com/czbiohub/sourmash/pull/11

## preparation:

build test1 and test2 SBTs:

```
sourmash index -k 21 test1 tests/test-data/gather/GCF_000011885.1_ASM1188v1_genomic.fna.gz.sig tests/test-data/gather/GCF_000016785.1_ASM1678v1_genomic.fna.gz.sig tests/test-data/gather/GCF_000006945.2_ASM694v2_genomic.fna.gz.sig tests/test-data/gather/GCF_000009085.1_ASM908v1_genomic.fna.gz.sig tests/test-data/gather/GCF_000016045.1_ASM1604v1_genomic.fna.gz.sig tests/test-data/gather/GCF_000007545.1_ASM754v1_genomic.fna.gz.sig tests/test-data/gather/GCF_000195995.1_ASM19599v1_genomic.fna.gz.sig tests/test-data/gather/GCF_000008545.1_ASM854v1_genomic.fna.gz.sig tests/test-data/gather/GCF_000009505.1_ASM950v1_genomic.fna.gz.sig tests/test-data/gather/GCF_000009525.1_ASM952v1_genomic.fna.gz.sig tests/test-data/gather/GCF_000008105.1_ASM810v1_genomic.fna.gz.sig tests/test-data/gather/GCF_000018945.1_ASM1894v1_genomic.fna.gz.sig
sourmash index -k 21 test2 tests/test-data/gather/GCF_000006945.2_ASM694v2_genomic.fna.gz.sig tests/test-data/gather/GCF_000007545.1_ASM754v1_genomic.fna.gz.sig tests/test-data/gather/GCF_000008105.1_ASM810v1_genomic.fna.gz.sig tests/test-data/gather/GCF_000008545.1_ASM854v1_genomic.fna.gz.sig tests/test-data/gather/GCF_000009085.1_ASM908v1_genomic.fna.gz.sig tests/test-data/gather/GCF_000009505.1_ASM950v1_genomic.fna.gz.sig tests/test-data/gather/GCF_000009525.1_ASM952v1_genomic.fna.gz.sig tests/test-data/gather/GCF_000011885.1_ASM1188v1_genomic.fna.gz.sig tests/test-data/gather/GCF_000016045.1_ASM1604v1_genomic.fna.gz.sig tests/test-data/gather/GCF_000016785.1_ASM1678v1_genomic.fna.gz.sig tests/test-data/gather/GCF_000018945.1_ASM1894v1_genomic.fna.gz.sig tests/test-data/gather/GCF_000195995.1_ASM19599v1_genomic.fna.gz.sig

```

Verify that they produce different numbers of results with same gather query (note: gather should return 12 results :)
```
# gather against test1 produces 11
sourmash gather tests/test-data/gather/combined.sig test1 --threshold-bp=0
# gather against test2 produces 8
sourmash gather tests/test-data/gather/combined.sig test2 --threshold-bp=0
```

save the unassigned bits from the test2 search:

```
sourmash gather tests/test-data/gather/combined.sig test2 --threshold-bp=0 --output-unassigned test2.un.sig
```

## dig into the test2 tree with the unassigned bits

In the below tree, the match score at the end of each line is the containment of the unassigned bits. I've hand annotated "things that shouldn't happen" with X1 and X2:

```
% ./print-sbt-3.py test2 test2.un.sig
 *Node:internal.0 [occupied: 984, fpr: 9.4e-09] match=0
X1 *Node:internal.2 [occupied: 1202, fpr: 2.1e-08] match=1
     *Node:internal.6 [occupied: 1058, fpr: 1.3e-08] match=1
       *Node:internal.14 [occupied: 0, fpr: 0.0] match=0
       *Node:internal.13 [occupied: 0, fpr: 0.0] match=0
     *Node:internal.5 [occupied: 1015, fpr: 1.1e-08] match=1
       *Node:internal.12 [occupied: 346, fpr: 1.4e-10] match=0
         **Leaf:NC_002163.1 Campylobacter jejuni  match=0
         **Leaf:NC_009486.1 Thermotoga petrophila match=0
       *Node:internal.11 [occupied: 772, fpr: 3.6e-09] match=1
         **Leaf:NC_006905.1 Salmonella enterica s match=0
         **Leaf:NC_006511.1 Salmonella enterica s match=1
X2 *Node:internal.1 [occupied: 1363, fpr: 3.5e-08] match=1
     *Node:internal.4 [occupied: 1315, fpr: 3e-08] match=1
       *Node:internal.10 [occupied: 565, fpr: 1e-09] match=1
         **Leaf:NC_004631.1 Salmonella enterica s match=1
         **Leaf:NC_003198.1 Salmonella enterica s match=0
       *Node:internal.9 [occupied: 517, fpr: 7.2e-10] match=0
         **Leaf:NC_011978.1 Thermotoga neapolitan match=0
         **Leaf:NC_000853.1 Thermotoga maritima M match=0
     *Node:internal.3 [occupied: 971, fpr: 8.9e-09] match=1
       *Node:internal.8 [occupied: 922, fpr: 7.2e-09] match=1
         **Leaf:NC_003197.2 Salmonella enterica s match=0
         **Leaf:NC_011080.1 Salmonella enterica s match=0
       *Node:internal.7 [occupied: 730, fpr: 2.8e-09] match=1
         **Leaf:NC_011274.1 Salmonella enterica s match=1
         **Leaf:NC_011294.1 Salmonella enterica s match=1
```

Details: here you can see that internal.2 and internal.1 are both children of internal.0, and both have matches; but internal.0 doesn't match! This should never happen in an SBT.

## next, dig into the test1 tree with its unassigned bit

First, get the unassigned bits remaining after gather against test1 --

```
sourmash gather test2.un.sig test1 --threshold=0 --output-unassigned test1.un.sig
```

Note that a search against signatures, rather than a tree,

```
sourmash gather test1.un.sig tests/test-data/gather/GCF*.sig --threshold=0
```
yields a match to `NC_004631.1`:
```
overlap     p_query p_match
---------   ------- -------
20.0 kbp     100.0%    0.4%    NC_004631.1 Salmonella enterica subsp...

found 1 matches total;
the recovered matches hit 100.0% of the query
```

When I look at the test1 tree:
```
./print-sbt-3.py test1 test1.un.sig 
 *Node:internal.0 [occupied: 786, fpr: 3.8e-09] match=0
   *Node:internal.2 [occupied: 1411, fpr: 4e-08] match=1
Y1   *Node:internal.6 [occupied: 1028, fpr: 1.1e-08] match=1
       *Node:internal.14 [occupied: 0, fpr: 0.0] match=0
       *Node:internal.13 [occupied: 0, fpr: 0.0] match=0
     *Node:internal.5 [occupied: 1326, fpr: 3.1e-08] match=0
       *Node:internal.12 [occupied: 656, fpr: 1.9e-09] match=0
         **Leaf:NC_006905.1 Salmonella enterica s match=0
         **Leaf:NC_011978.1 Thermotoga neapolitan match=0
       *Node:internal.11 [occupied: 575, fpr: 1.1e-09] match=0
         **Leaf:NC_011080.1 Salmonella enterica s match=0
         **Leaf:NC_003197.2 Salmonella enterica s match=0
=> *Node:internal.1 [occupied: 1234, fpr: 2.3e-08] match=1
=>   *Node:internal.4 [occupied: 1187, fpr: 2e-08] match=1
       *Node:internal.10 [occupied: 581, fpr: 1.1e-09] match=0
         **Leaf:NC_006511.1 Salmonella enterica s match=0
         **Leaf:NC_002163.1 Campylobacter jejuni  match=0
=>     *Node:internal.9 [occupied: 565, fpr: 1e-09] match=1
         **Leaf:NC_003198.1 Salmonella enterica s match=0
=>       **Leaf:NC_004631.1 Salmonella enterica s match=1
     *Node:internal.3 [occupied: 1148, fpr: 1.7e-08] match=0
       *Node:internal.8 [occupied: 443, fpr: 3.9e-10] match=0
         **Leaf:NC_009486.1 Thermotoga petrophila match=0
         **Leaf:NC_000853.1 Thermotoga maritima M match=0
       *Node:internal.7 [occupied: 897, fpr: 6.5e-09] match=0
         **Leaf:NC_011294.1 Salmonella enterica s match=0
         **Leaf:NC_011274.1 Salmonella enterica s match=0
```
You can see that again the root note has no matches, despite both internal nodes having matches (presumably the Y1 is a false positive? that's at least plausible :)

## conclusion

the observed problem is that the root node is not getting properly updated with contents beneath it. I don't know if that's the whole problem, though! HTH.

**Comment by ctb** [comment]
makes sense! semi-random thought: I wonder if the problem is that the bloom filters don't get updated when the tree structure is changed? I was looking through the code and it looked to me like most of the time if a Nodegraph existed it was no longer updated again. But I couldn't nail that down before I ran out of time and had to go look at something else.

**Comment by ctb** [comment]
(this could lead to both observed problems, because once the intermediate node bloom filters are out of date, the placement of new nodes becomes wrong too)
